src/modals/storage-connections/StorageConnectionsTable.js

```diff
@@ -0,0 +1,340 @@
+  }
+
+  const checkFieldsEdited = (row) => {
+    const rowEdited = row.cells
+      .map(cell => cell.props.value !== cell.props.editableValue)
+      .reduce((acc, edited) => acc || edited)
+    return rowEdited
+  }
+
+  const updateRows = (evt, type, isEditable, rowIndex, validationErrs) => {
+    const newRows = Array.from(rows)
+    const connection = newRows[rowIndex].connection
+
+    switch (type) {
+      case 'edit':
+        replaceAttachedCellWithEmptyCell(newRows[rowIndex])
+        newRows[rowIndex] = applyCellEdits(newRows[rowIndex], type)
+        setRows(newRows)
+        break
+      case 'cancel':
+        if (isNewConnectionOpened && rowIndex === 0) {
+          setNewConnectionOpened(false)
+        } else {
+          newRows[rowIndex] = cancelCellEdits(newRows[rowIndex])
+          replaceAttachedCellWithIconCell(newRows[rowIndex], connection)
+          setRows(newRows)
+        }
+        break
+      case 'save': {
+        // Replacing the default PatternFly Table input validation mechanism
+        const validationErrors = validateRow(newRows[rowIndex])
+        if (validationErrors && Object.keys(validationErrors).length) {
+          newRows[rowIndex] = validateCellEdits(newRows[rowIndex], type, validationErrors)
+          setRows(newRows)
+        } else {
+          const newConnection = getConnectionFromRow(newRows[rowIndex])
+          if (isNewConnectionOpened && rowIndex === 0) {
+            setNewConnectionOpened(false)
+            onCreate(newConnection)
+          } else if (checkFieldsEdited(newRows[rowIndex])) {
+            onEdit(newConnection, connection.id)
+          } else {
+            newRows[rowIndex] = cancelCellEdits(newRows[rowIndex])
+            replaceAttachedCellWithIconCell(newRows[rowIndex], connection)
+            setRows(newRows)
+          }
+        }
+        break
+      }
+      default:
+        break
+    }
+  }
+
+  // This function replaces the regular PF Table input validation mechanism in Editable Cells, which
+  // as for now doesn't have an option to apply rules on separate cells, but rather only on all the cells in a row
+  // This function validates every cell according to cell.props.rules
+  // returns map (rule name => array of cell names that failed validation)
+  const validateRow = (row) => {
+    const validationErrors = row.cells.reduce((errorsMap, cell, index) => {
+      const testValue = cell.props.editableValue === '' ? '' : cell.props.editableValue || cell.props.value
+      const cellErrors = cell.props.rules.reduce((failedRules, rule) => {
+        if (!rule.validator(testValue)) {
+          failedRules.push(rule.name)
+        }
+        return failedRules
+      }, [])
+      cellErrors.forEach(ruleName => {
+        if (!errorsMap[ruleName]) {
+          errorsMap[ruleName] = []
+        }
+        errorsMap[ruleName].push(cell.props.name)
+      })
+      return errorsMap
+    }, {})
+    return validationErrors
```

Comment:
@sjd78 I agree that `reduce` is not always the best choice but in case it could work but fully functional flow needs to be used i.e.
```
row.cells
      .filter(({ editableValue }) => /* hasEditableValue */ true)
      .flatMap(({ name, editableValue, rules }) =>
        rules.filter(rule => !rule.validator(editableValue))
          .map(rule => ({ cell: name, rule: rule.name }))
      ).reduce((acc, { cell, rule }) => ({ ...acc, rule: [...acc[rule], cell] }), {})

```

Reply:
@rszwajko - My cognative load to read the original is as high as to read yours.  I like the for loops just to keep it simple.